tests/eval/test_2fa_duo.py

```diff
@@ -0,0 +1,408 @@
+"""
+This set of tests is based on the manual tests found in the
+IdP Release Candidate Test Matrix Version 4.2.4 (5/27/2021) here
+https://wiki.cac.washington.edu/display/SMW/IAM+Team+Wiki
+
+2FA-1 thru 2FA-11. 2FA-8b and 2FA-10 are not yet automate-able.
+"""
+
+import pytest
+
+from webdriver_recorder.browser import Chrome
+
+from tests.helpers import CTRL_KEY, Locators, WebTestUtils
+from tests.models import ServiceProviderInstance, AccountNetid, TestSecrets
+
+from selenium import webdriver
+
+
+@pytest.fixture
+def browser() -> Chrome:
+    return Chrome()
+
+
+@pytest.fixture(scope='class')
+def class_browser() -> Chrome:
+    browser = Chrome()
+    try:
+        yield browser
+    finally:  # Even if there is an error
+        browser.close()
+
+
+@pytest.fixture
+def new_tab(browser):
+    def inner():
+        browser.execute_script("window.open('');")
+        browser.switch_to.window(browser.window_handles[-1])
+    return inner
+
+
+@pytest.fixture
+def close_tab(browser):
+    def inner():
+        browser.find_element_by_tag_name('body').send_keys(CTRL_KEY + 'w')
+    return inner
+
+
+@pytest.fixture
+def netid() -> str:
+    return AccountNetid.sptest01.value
+
+
+@pytest.fixture
+def netid2() -> str:
+    return AccountNetid.sptest02.value
+
+
+@pytest.fixture
+def netid3() -> str:
+    return AccountNetid.sptest03.value
+
+
+@pytest.fixture
+def netid4() -> str:
+    return AccountNetid.sptest04.value
+
+
+@pytest.fixture
+def netid5() -> str:
+    return AccountNetid.sptest05.value
+
+
+@pytest.fixture
+def clean_browser(browser: Chrome) -> Chrome:
+    browser.delete_all_cookies()
+    return browser
+
+
+def add_suffix(suffix: str, netid: str) -> str:
+    return f'{netid}{suffix}'
+
+
+def crazy_casing(netid: str) -> str:
```

Comment:
re: `crazy_casing`: Please avoid copy-pasting code like this, just factor the code so that it can be imported from a single place. I recommend creating `helpers.py` and adding the `crazy_casing` function there, or re-defining it as a fixture (in the root `conftest.py`) so it can be used from any test everywhere without importing.

One reason this is a problem in this particular instance is because the function name should really be changed to avoid use of the word `crazy` (We've learned new information since the time I wrote these tests); by adding a _new_ function here, you are responsible for changing the function name in _this_ test to something else, which would mean we have two functions with different names doing the same thing in the same code base.

If you factor and import it, you are not responsible for changing the name, because you aren't introducing a new symbol into the code base, just moving an existing one, so we can scope that change separately. :) 

NB: The function name was my original comment, before remembering that it sounded familiar and may have come from me. :)

Reply:
Yes, `crazy_casing` was inherited from earlier versions of this test suite.
I renamed the function to `lower_upper_casing` and tried putting it in `confest.py` but I keep getting the error:

TypeError: 'str' object is not callable
when I do something like...
`crazy_netid = self.lower_upper_casing(netid)`